File: packages/react-native-nitro-test/ios/HybridRecyclableTestView.swift
//
//  HybridRecyclableTestView.swift
//  react-native-nitro-test
//
//  Created by Marc Rousavy on 03.10.24.
//

import NitroModules
import UIKit

class HybridRecyclableTestView: HybridRecyclableTestViewSpec, RecyclableView {
  // UIView
  var view: UIView = UIView()
  private let lifecycleLock = NSLock()
  private var isRecycled = false
  private var invalidLifecycleOrderCount: Double = 0
  private var onDropViewCount: Double = 0
  private var prepareForRecycleCount: Double = 0
  private var nativeDefaultValueSetterCallCount: Double = 0
  private var nativeDefaultValueStorage: Double? = 42

  // Props
  var isBlue: Bool = false {
    didSet {
      if !isRecycled {
        view.backgroundColor = isBlue ? .systemBlue : .systemRed
      }
    }
  }
  var nativeDefaultValue: Double? {
    get {
      return nativeDefaultValueStorage
    }
    set {
      nativeDefaultValueStorage = newValue
      nativeDefaultValueSetterCallCount += 1
    }
  }

  func onDropView() {
    print("View dropped!")
    lifecycleLock.withLock {
      onDropViewCount += 1
    }
  }

  func getOnDropViewCount() throws -> Double {
    return lifecycleLock.withLock { onDropViewCount }
  }

  func getInvalidLifecycleOrderCount() throws -> Double {
    return lifecycleLock.withLock { invalidLifecycleOrderCount }
  }

  func getPrepareForRecycleCount() throws -> Double {
    return lifecycleLock.withLock { prepareForRecycleCount }
  }

  func getNativeDefaultValueSetterCallCount() throws -> Double {
    return nativeDefaultValueSetterCallCount
  }

  func beforeUpdate() {
    isRecycled = false
  }

  // Recycling conformance
  func prepareForRecycle() {
    nativeDefaultValueStorage = 42
    nativeDefaultValueSetterCallCount = 0
    view.backgroundColor = .yellow
    isRecycled = true

    lifecycleLock.withLock {
      if onDropViewCount != prepareForRecycleCount + 1 {
        invalidLifecycleOrderCount += 1
      }
      prepareForRecycleCount += 1
    }
  }

}

File: packages/react-native-nitro-test/android/src/main/java/com/margelo/nitro/test/HybridRecyclableTestView.kt
package com.margelo.nitro.test

import android.graphics.Color
import android.util.Log
import android.view.View
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.views.RecyclableView
import java.util.concurrent.atomic.AtomicInteger

@Keep
@DoNotStrip
class HybridRecyclableTestView(
  val context: ThemedReactContext,
) : HybridRecyclableTestViewSpec(),
  RecyclableView {
  // View
  override val view: View = View(context)
  private var isRecycled = false
  private val invalidLifecycleOrderCount = AtomicInteger()
  private val onDropViewCount = AtomicInteger()
  private val prepareForRecycleCount = AtomicInteger()
  private var nativeDefaultValueSetterCallCount = 0.0
  private var nativeDefaultValueStorage: Double? = 42.0

  // Props
  override var isBlue: Boolean = false
    set(value) {
      field = value
      if (!isRecycled) {
        val color = if (value) Color.BLUE else Color.RED
        view.setBackgroundColor(color)
      }
    }
  override var nativeDefaultValue: Double?
    get() = nativeDefaultValueStorage
    set(value) {
      nativeDefaultValueStorage = value
      nativeDefaultValueSetterCallCount += 1
    }

  override fun onDropView() {
    Log.i(TAG, "View dropped!")
    onDropViewCount.incrementAndGet()
  }

  override fun getOnDropViewCount(): Double = onDropViewCount.get().toDouble()

  override fun getInvalidLifecycleOrderCount(): Double = invalidLifecycleOrderCount.get().toDouble()

  override fun getPrepareForRecycleCount(): Double = prepareForRecycleCount.get().toDouble()

  override fun getNativeDefaultValueSetterCallCount(): Double = nativeDefaultValueSetterCallCount

  override fun beforeUpdate() {
    isRecycled = false
  }

  // Recycling conformance
  override fun prepareForRecycle() {
    if (onDropViewCount.get() != prepareForRecycleCount.get() + 1) {
      invalidLifecycleOrderCount.incrementAndGet()
    }
    nativeDefaultValueStorage = 42.0
    nativeDefaultValueSetterCallCount = 0.0
    view.setBackgroundColor(Color.YELLOW)
    isRecycled = true
    prepareForRecycleCount.incrementAndGet()
  }
}
